website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+            } else if (response.status === 'error') {
+                self.onRegisterFail();
+            }
+        }).always($osf.unblock).fail(self.onRegisterFail);
+};
+
+
+/**
+ * @class RegistrationEditor
+ *
+ * @param {Object} urls
+ * @param {String} urls.update: endpoint to update a draft instance
+ * @param {String} editorID: id of editor DOM node
+ * @property {ko.observable[Boolean]} readonly
+ * @property {ko.observable[Draft]} draft
+ * @property {ko.observable[Question]} currentQuestion
+ * @property {Object} extensions: mapping of extenstion names to their view models
+ **/
+var RegistrationEditor = function(urls, editorId) {
+
+    var self = this;
+
+    self.urls = urls;
+
+    self.readonly = ko.observable(false);
+
+    self.draft = ko.observable();
+
+    self.currentQuestion = ko.observable();
+    // When the currentQuestion changes, save when it's rate-limited value changes
+    self.currentQuestion.subscribe(function(question) {
+         question.delayedValue.subscribe(self.save.bind(self));
+    });
+    self.showValidation = ko.observable(false);
+
+    self.currentPages = ko.computed(function() {
+        var draft = self.draft();
+        if (!draft) {
+            return [];
+        }
+        var schema = draft.schema();
+        if (!schema) {
+            return [];
+        }
+        return schema.pages;
+    });
+
+    self.lastSaveTime = ko.computed(function() {
+        if (!self.draft()) {
+            return null;
+        }
+        return self.draft().updated;
+    });
+    self.lastSaved = ko.computed(function() {
+        var t = self.lastSaveTime();
+        if (t) {
+            return t.toGMTString();
+        } else {
+            return 'never';
+        }
+    });
+
+    self.canRegister = ko.computed(function() {
+        var draft = self.draft();
+        return draft && draft.isApproved;
+    });
+
+    self.iterObject = $osf.iterObject;
+
+    // TODO: better extensions system?
+    self.extensions = {
+        'osf-upload': editorExtensions.Uploader
+    };
+};
+/**
+ * Load draft data into the editor
+ *
+ * @param {Draft} draft
+ **/
+RegistrationEditor.prototype.init = function(draft) {
+    var self = this;
+
+    self.draft(draft);
+    var metaSchema = draft.metaSchema;
+
+    var schemaData = {};
+    if (draft) {
+        schemaData = draft.schemaData || {};
+    }
+
+    var questions = self.flatQuestions();
+    $.each(questions, function(i, question) {
+        var val = schemaData[question.id];
+        if (val) {
+            if (question.type === 'object') {
+                $.each(question.properties, function(prop, subQuestion) {
+                    val = schemaData[question.id][prop];
+                    if (val) {
+                        subQuestion.value(val.value);
+                    }
+                });
+            } else {
+                question.value(val.value);
+            }
+        }
+    });
+    self.currentQuestion(questions.shift());
```

Comment:
There's a lot of intermediate transformation of data here; Lyndsy has mentioned that he's planning a refactor, so I'll do a light review of this section until his changes are finished.